### PR TITLE
Add support for Resources in the SDK

### LINKED ISSUE #515: Default SDK support for resources (Metrics)
There is  interest in specifying a standard environment variable for passing resources in the default SDKs. See https://github.com/open-telemetry/opentelemetry-specification/issues/501. Although we do not have a specification for a particular environment variable name, we can assume that there will be resources configured in the SDKs and that exporters will have access to them.

Resource support should be added in the Metric SDK.

### PR DESCRIPTION
**Pull request by MrAlias**
Add `Config` types for the push `Controller` and the `SDK`. Included with this are helper functions to configure the `ErrorHandler` and `Resource`.

Add a `Resource` to the `Descriptor`. The choice to add the `Resource` here (instead of say a `Record` or the `Instrument` itself) was motivated by the definition of the `Descriptor` as the way to
uniquely describe a metric instrument.

Update the push `Controller` and default `SDK` to pass down their configured `Resource` from instantiation to the metric instruments.

Resolves #515

### CODE REVIEW

**Comment by rghetia** [comment]
> Add a `Resource` to the Meter `Descriptor`. The choice to add the `Resource` here (instead of say a `Record` or the `Instrument` itself) was motivated by the definition of the `Descriptor` as the way to
> uniquely describe a metric instrument.
> 

How about WithResource as an option? I guess same could be applied to unit, description, and labels as well.

File: sdk/metric/controller/push/push.go [hunk 1]
```diff
@@ -73,16 +73,21 @@ var _ Ticker = realTicker{}
 // exporter will be used as the label encoder for the SDK itself,
 // otherwise the SDK will be configured with the default label
 // encoder.
-func New(batcher export.Batcher, exporter export.Exporter, period time.Duration) *Controller {
+func New(batcher export.Batcher, exporter export.Exporter, period time.Duration, opts ...Option) *Controller {
```
Comment (Aneurysm9) [inline sdk/metric/controller/push/push.go]: Should the documentation here be updated to reflect that this function now accepts `Option`s?

Comment (MrAlias) [reply sdk/metric/controller/push/push.go]: Updated

**Comment by MrAlias** [comment]
> How about WithResource as an option? I guess same could be applied to unit, description, and labels as well.

I think this is a good direction. It would follow our other design patterns as well. I'll refactor.

File: sdk/export/metric/descriptor.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+import (
+	"go.opentelemetry.io/otel/api/core"
+	"go.opentelemetry.io/otel/api/unit"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+// DescriptorConfig contains configuration for a metric Descriptor
+type DescriptorConfig struct {
+	// keys are the common keys related to all measurements.
```
Comment (rghetia) [inline sdk/export/metric/descriptor.go]:
```suggestion
	// Keys are the common keys related to all measurements.
```

File: sdk/export/metric/descriptor.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	Keys []core.Key
+
+	// description describes the metric in human readable terms.
```
Comment (rghetia) [inline sdk/export/metric/descriptor.go]:
```suggestion
	// Description describes the metric in human readable terms.
```

File: sdk/export/metric/descriptor.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	Description string
+
+	// unit is the determinate quantity used as a standard of all
```
Comment (rghetia) [inline sdk/export/metric/descriptor.go]:
```suggestion
	// Unit is the determinate quantity used as a standard of all
```

File: sdk/export/metric/descriptor.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	// measurements for the metric.
+	Unit unit.Unit
+
+	// resource descrbes the entity for which this metric makes measurements.
```
Comment (rghetia) [inline sdk/export/metric/descriptor.go]:
```suggestion
	// Resource describes the entity for which this metric makes measurements.
```

**Comment by jmacd** [comment]
I have a selfish favor to ask. Would you mind waiting until #560 merges? (If it merges.)

This PR will cause a lot of conflicts with #560, which I'm nearly finished with. #560 was a really difficult refactoring, and I'm worried about conflicts. The cause of the conflict is that I needed a Descriptor type in the API to organize the two new constructors (NewSynchronousInstrument, ...), so I removed `export.Descriptor` and moved it into `api/metric`. I suspect you'll have a much easier time resolving these conflicts in here than I will have in #560. 

I am self-reviewing #560 now and adding comments and cleanups, but it's building and passing tests.

**Comment by MrAlias** [comment]
> Would you mind waiting until #560 merges?

Yeah, no problem.

File: sdk/export/metric/descriptor.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	// Unit is the determinate quantity used as a standard of all
+	// measurements for the metric.
+	Unit unit.Unit
+
+	// Resource describes the entity for which this metric makes measurements.
+	Resource resource.Resource
+}
+
+// DescriptorOption is the interface that applies the value to a DescriptorConfig option.
+type DescriptorOption interface {
```
Comment (krnowak) [inline sdk/export/metric/descriptor.go]: We usually had a convention of option type being something like `type DescriptionOption func(*DescriptorConfig)`. Is there a reason for being different here?

File: sdk/export/metric/descriptor.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+// WithKeys applies common label keys.
+// Multiple `WithKeys` options accumulate.
+func WithKeys(keys ...core.Key) DescriptorOption {
+	return keysOption(keys)
+}
+
+type keysOption []core.Key
+
+func (k keysOption) Apply(config *DescriptorConfig) {
+	if config == nil {
```
Comment (krnowak) [inline sdk/export/metric/descriptor.go]: Should we even care about `nil` config? `NewDescriptor` never passes `nil` and I don't think these options are going to be used by anyone else.

File: sdk/metric/config.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	// with the `SetErrorHandler` method.
+	ErrorHandler ErrorHandler
+
+	// Resource is the OpenTelemetry resource associated with all Meters
+	// created by the SDK.
+	Resource resource.Resource
+}
+
+// Option is the interface that applies the value to a configuration option.
+type Option interface {
```
Comment (krnowak) [inline sdk/metric/config.go]: Same here about interface vs function.

File: sdk/metric/config.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+// WithErrorHandler sets the ErrorHandler configuration option of a Config.
+func WithErrorHandler(fn ErrorHandler) Option {
+	return errorHandlerOption(fn)
+}
+
+type errorHandlerOption ErrorHandler
+
+func (o errorHandlerOption) Apply(config *Config) {
+	if config == nil {
```
Comment (krnowak) [inline sdk/metric/config.go]: Same here about `nil` config.

File: sdk/metric/controller/push/config.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	// with the `SetErrorHandler` method.
+	ErrorHandler sdk.ErrorHandler
+
+	// Resource is the OpenTelemetry resource associated with all Meters
+	// created by the Controller.
+	Resource resource.Resource
+}
+
+// Option is the interface that applies the value to a configuration option.
+type Option interface {
```
Comment (krnowak) [inline sdk/metric/controller/push/config.go]: Same here about interface vs function.

File: sdk/metric/controller/push/config.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+// WithErrorHandler sets the ErrorHandler configuration option of a Config.
+func WithErrorHandler(fn sdk.ErrorHandler) Option {
+	return errorHandlerOption(fn)
+}
+
+type errorHandlerOption sdk.ErrorHandler
+
+func (o errorHandlerOption) Apply(config *Config) {
+	if config == nil {
```
Comment (krnowak) [inline sdk/metric/controller/push/config.go]: Same here about `nil` config.

File: sdk/metric/controller/push/push.go [hunk 12]
```diff
@@ -65,24 +65,29 @@ var _ Clock = realClock{}
...
 
+	c := &Config{ErrorHandler: sdk.DefaultErrorHandler}
+	for _, opt := range opts {
+		opt.Apply(c)
+	}
+
 	return &Controller{
-		sdk:          sdk.New(batcher, lencoder),
-		errorHandler: sdk.DefaultErrorHandler,
+		sdk:          sdk.New(batcher, lencoder, sdk.WithResource(c.Resource)),
```
Comment (krnowak) [inline sdk/metric/controller/push/push.go]: Should we also forward the error handler to SDK here too?

**Comment by krnowak** [comment]
I think that the discussion about option as interface vs function can be moved to #536, so no need to block this PR on it. Still there was one issue about forwarding the error handler to address.

File: sdk/export/metric/descriptor.go (hunk 6 above)
Comment (MrAlias) [reply sdk/export/metric/descriptor.go]: This is a good point. Plan is to address with #536.

File: sdk/metric/config.go (hunk 8 above)
Comment (MrAlias) [reply sdk/metric/config.go]: Same, plan is to address with #536.

File: sdk/metric/config.go (hunk 9 above)
Comment (MrAlias) [reply sdk/metric/config.go]: Good point, will fix.

File: sdk/metric/controller/push/config.go (hunk 10 above)
Comment (MrAlias) [reply sdk/metric/controller/push/config.go]: Same, plan is to address with #536.

File: sdk/metric/controller/push/config.go (hunk 11 above)
Comment (MrAlias) [reply sdk/metric/controller/push/config.go]: will fix.

File: sdk/metric/controller/push/push.go (hunk 12 above)
Comment (MrAlias) [reply sdk/metric/controller/push/push.go]: Good catch. Updated.

**Comment by MrAlias** [comment]
I re-requested review from you @krnowak and @rghetia due to how much this changed in strategy now that #560 was merged. PTAL

File: sdk/metric/sdk.go [hunk 13]
```diff
@@ -559,6 +570,16 @@ func (m *SDK) checkpoint(ctx context.Context, descriptor *metric.Descriptor, rec
 	return 1
 }
 
+// Resource returns the Resource this SDK was created with describing the
+// entity for which it creates instruments for.
+//
+// Resource means that the SDK implements the Resourcer interface and
+// therefore all metric instruments it creates will inherited its
```
Comment (rghetia) [inline sdk/metric/sdk.go]:
```suggestion
// therefore all metric instruments it creates will inherit its
```

File: sdk/resource/resource.go [hunk 14]
```diff
@@ -70,3 +72,8 @@ func (r Resource) Attributes() []core.KeyValue {
 	}
 	return attrs
 }
+
+// Equal returns if other Resource is the equal to r.
```
Comment (rghetia) [inline sdk/resource/resource.go]:
```suggestion
// Equal returns true if other Resource is the equal to r.
```

**Pull request merged**